File: python/hash_table.py
def create_hash(word: str):
    return len(word)


# O(n * t), 'n' being length of values, 't' being length of hash_table
def populate_hash_table(hash_table, values):
    for value in values:
        hash_value = create_hash(value)

        if hash_value in hash_table: hash_table[hash_value].append(value)
        else: hash_table[hash_value] = [value]


# O(n), 'n' being length of hash_table[hash_value], O(1) in real life
def does_element_exist_in_hash_table(hash_table, value):
    hash_value = create_hash(value)

    if not hash_value in hash_table: return False

    return value in hash_table[hash_value]


if __name__ == '__main__':
    hash_table = dict()

    names = [ 'Dr4kk0nnys', 'Bob', 'Alex', 'Creed', 'Ana', 'Joan', 'Timmy' ]

    # v1
    for name in names:
        hash_table[name] = 'Some important value ' + name

    # constant lookup time. Output: 'Some important value Dr4kk0nnys'
    print(hash_table['Dr4kk0nnys'])


    # v2
    hash_table = dict()
    populate_hash_table(hash_table, names)

    # O(n), 'n' being the size of hash_table[create_hash('Ana')]. In real life, is constant lookup time
    print(does_element_exist_in_hash_table(hash_table, 'Ana'))
    print(hash_table)

File: typescript/src/hashTable.ts
class HashTable {
    
    private hashTable = {};

    private createHash = (value: any) => value[0] + String(value).length;

    insert = (values: any[] | any) => {

        if (!Array.isArray(values)) values = [values];

        for (const value of values) {
            const hashValue = String(this.createHash(value));

            if (this.hashTable[hashValue]) { this.hashTable[hashValue].push(value); }
            else { this.hashTable[hashValue] = [value]; }
        }
    }

    // O(1) in real life.
    includes = (value: any) => {
        const hashValue = String(this.createHash(value));

        if (!this.hashTable[hashValue]) return false;

        return this.hashTable[hashValue].includes(value);
    }

    logHashTable = () => console.log(this.hashTable);
}


const hashTable = new HashTable();
const names = [ 'Dr4kk0nnys', 'Bob', 'Alex', 'Creed', 'Anna', 'Joan', 'Timmy' ];

hashTable.insert(names);
hashTable.insert('Dr4kk');
hashTable.logHashTable();

console.log(hashTable.includes('Bob')); // true
console.log(hashTable.includes('Alex')); // true
console.log(hashTable.includes('Bobs')); // false
